File: transaction-service/src/main/java/ua/com/foxminded/restClient/service/TransactionServiceImpl.java
package ua.com.foxminded.restClient.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import ua.com.foxminded.restClient.dto.TransactionDto;
import ua.com.foxminded.restClient.entity.Transaction;
import ua.com.foxminded.restClient.exceptions.PersonNotFoundException;
import ua.com.foxminded.restClient.mapper.TransactionMapper;
import ua.com.foxminded.restClient.repository.TransactionRepository;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
public class TransactionServiceImpl implements TransactionService {
    private TransactionMapper mapper;
    private TransactionRepository repository;

    @Autowired
    public TransactionServiceImpl(TransactionMapper mapper, TransactionRepository repository) {
        this.mapper = mapper;
        this.repository = repository;
    }

    @Override
    public Page<TransactionDto> findAllByIdAndBetweenDate(UUID id, LocalDateTime start, LocalDateTime end, Pageable pageable) {
        Page<Transaction> transactions = repository.findAllByPersonIdAndTransactionTimeBetween(id, start, end, pageable);
        if (transactions.isEmpty()) {
            throw new PersonNotFoundException(id);
        }
        return transactions.map(mapper::entityToDto);
    }

}

File: transaction-service/src/main/kotlin/ua/com/foxminded/restClient/service/TransactionServiceImpl.kt
package ua.com.foxminded.restClient.service

import java.time.LocalDateTime
import java.util.UUID
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import ua.com.foxminded.restClient.dto.TransactionDto
import ua.com.foxminded.restClient.exceptions.PersonNotFoundException
import ua.com.foxminded.restClient.mapper.TransactionMapper
import ua.com.foxminded.restClient.repository.TransactionRepository


@Service
class TransactionServiceImpl(
    private val mapper: TransactionMapper,
    private val repository: TransactionRepository
) : TransactionService {

    override fun findAllByIdAndBetweenDate(
        id: UUID,
        start: LocalDateTime?,
        end: LocalDateTime?,
        pageable: Pageable
    ): Flux<TransactionDto> {
        val endDate = when {
            end == null -> LocalDateTime.now()
            end.isAfter(start) -> LocalDateTime.now()
            else -> end
        }
        return repository
            .findAllByPersonIdAndTransactionTimeBetween(id, start!!, endDate, pageable)
            .switchIfEmpty(Flux.error(PersonNotFoundException(id)))
            .map { mapper.entityToDto(it) }

    }
}
